tests/migrations/test_autodetector.py

```diff
@@ -1369,6 +1370,37 @@ def test_add_date_fields_with_auto_now_add_asking_for_default(
         self.assertOperationFieldAttributes(changes, "testapp", 0, 2, auto_now_add=True)
         self.assertEqual(mocked_ask_method.call_count, 3)
 
+    def test_add_field_before_generated_field(self):
+        initial_state = ModelState(
+            "testapp",
+            "Author",
+            [
+                ("name", models.CharField(max_length=20)),
+            ],
+        )
+        updated_state = ModelState(
+            "testapp",
+            "Author",
+            [
+                ("name", models.CharField(max_length=20)),
+                ("surname", models.CharField(max_length=20)),
+                (
+                    "lower_full_name",
+                    models.GeneratedField(
+                        expression=Concat(Lower("name"), Lower("surname")),
+                        output_field=models.CharField(max_length=30),
+                        db_persist=True,
+                    ),
+                ),
+            ],
+        )
+        changes = self.get_changes([initial_state], [updated_state])
+        self.assertNumberMigrations(changes, "testapp", 1)
+        self.assertOperationTypes(changes, "testapp", 0, ["AddField", "AddField"])
+        self.assertOperationFieldAttributes(
+            changes, "testapp", 0, 1, expression=Concat(Lower("name"), Lower("surname"))
+        )
+
```

Comment:
I was trying to think of a case that might fail during foreign keys additions along with a generated field , do you have one in mind? 🤔

Reply:
@DevilsAutumn 

1. Add a new model (e.g. `Foo`)
2. Add a foreign key to this model (`foo = models.ForeignKey("Foo", models.CASCADE)`)
3. Add a generated field that uses that attribute / column name of the foreign key (`foo_id_str = GeneratedField(Cast("foo_id", models.TextField()), ...)`.

It might already be working but my concerns here are that

1. Foreign keys are usually added last
2. Do we have logic in place so that `foo_id` is correctly mapped to the `foo` field addition by `attname`

Reply:
@charettes Thanks 👍 
Adding a fk along with a generatedfield was failing, I have fixed it and added a test case for the same.

I investigated for the `foo_id` mapping to `foo` as well and it seems to work as expected, Do you think its worth adding a test for this case as well?

Reply:
The tests added looks great @DevilsAutumn thanks for that!